lib/assertions.js

```diff
@@ -344,9 +344,100 @@ module.exports = expect => {
           this.outdentLines();
         });
       }
-      propertyNames.forEach(propertyName => {
-        expect(subject, '[not] to have [own] property', String(propertyName));
-      });
+
+      if (expect.flags.only) {
+        if (expect.flags.not) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "not" flag cannot be used together with "to only have properties".'
+          );
+        }
+        if (expect.flags.own) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "own" flag cannot be used together with "to only have properties".'
+          );
+        }
+        const subjectType = expect.subjectType;
+        const subjectKeys = subjectType.getKeys(subject).filter(
+          key =>
+            utils.numericalRegExp.test(key) ||
+            typeof key === 'symbol' ||
```

Comment:
All tests still pass if I comment out the two above lines, are they necessary? It also feels wrong that a numerical or symbol key with a value of `undefined` would be required?

```js
expect({ 123: undefined }, 'to only have properties', ['123']); // passes?!
```

If I actually pass in a symbol, an error is encountered because the numerical case is tested before the symbol one:

```js
const mySymbol = Symbol('foo');
expect({ [mySymbol]: undefined }, 'to only have properties', []);
```

```
/home/andreas/work/unexpected/lib/createTopLevelExpect.js:1514
    throw e;
    ^

TypeError: Cannot convert a Symbol value to a string
    at RegExp.test (<anonymous>)
    at subjectType.getKeys.filter.key (/home/andreas/work/unexpected/lib/assertions.js:364:35)
```

If I pass a symbol in the expected list of property names, I get:

```
expected { [Symbol('foo')]: undefined } to only have properties [ Symbol('foo') ]
  All expected properties must be passed as strings or numbers, but these are not:
    Symbol('foo')
```

... I guess that's a sign that there's still a bit of work to do before we support asserting on symbols everywhere, which might be a concern for a different PR.

Reply:
@papandreou yep, you've hit something. Re symbols we made "to satisfy" work via getKeys() but the port of this code from "to only have keys" seems to inherited its dislike of symbols. Let me look into this and undefined thing - worst case I agree we leave symbols out :)

Reply:
This was a good catch - those lines were were necessary elsewhere when whittling down the list of keys, but here we're only interested in those that are `undefined`. Incorporated the example above as a test case: https://github.com/unexpectedjs/unexpected/pull/671/files#diff-22b57e35705626d24e33c15ccf1c72e6R369